File: paper/benchmarks/bench_compact.py
"""Benchmark AIP compact token creation and verification (Python)."""
import sys, os, time, json, statistics
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'python'))

from aip_core.crypto import KeyPair
from aip_token.claims import AipClaims
from aip_token.compact import CompactToken

ITERATIONS = 1000

kp = KeyPair.generate()
claims = AipClaims(
    iss="aip:web:bench.test/agent",
    sub="aip:web:bench.test/tool",
    scope=["tool:search", "tool:browse"],
    budget_usd=1.0,
    max_depth=0,
    iat=int(time.time()),
    exp=int(time.time()) + 3600,
)

# Measure creation
create_times = []
for _ in range(ITERATIONS):
    start = time.perf_counter_ns()
    token = CompactToken.create(claims, kp)
    create_times.append((time.perf_counter_ns() - start) / 1_000_000)

# Measure verification
verify_times = []
pubkey = kp.public_key_bytes()
for _ in range(ITERATIONS):
    start = time.perf_counter_ns()
    CompactToken.verify(token, pubkey)
    verify_times.append((time.perf_counter_ns() - start) / 1_000_000)

token_size = len(token.encode('utf-8'))

results = {
    "language": "python",
    "mode": "compact",
    "iterations": ITERATIONS,
    "create_mean_ms": round(statistics.mean(create_times), 4),
    "create_p50_ms": round(statistics.median(create_times), 4),
    "create_p99_ms": round(sorted(create_times)[int(0.99 * ITERATIONS)], 4),
    "verify_mean_ms": round(statistics.mean(verify_times), 4),
    "verify_p50_ms": round(statistics.median(verify_times), 4),
    "verify_p99_ms": round(sorted(verify_times)[int(0.99 * ITERATIONS)], 4),
    "token_size_bytes": token_size,
}
print(json.dumps(results, indent=2))

outdir = os.path.join(os.path.dirname(__file__), 'results')
os.makedirs(outdir, exist_ok=True)
with open(os.path.join(outdir, 'compact_python.json'), 'w') as f:
    json.dump(results, f, indent=2)

File: rust/aip-token/src/bin/bench_compact.rs
//! Benchmark AIP compact token creation and verification (Rust).

use aip_core::crypto::KeyPair;
use aip_token::claims::AipClaims;
use aip_token::compact::CompactToken;
use std::time::Instant;

fn main() {
    let iterations = 1000;
    let kp = KeyPair::generate();
    let claims = AipClaims {
        iss: "aip:web:bench.test/agent".into(),
        sub: "aip:web:bench.test/tool".into(),
        scope: vec!["tool:search".into(), "tool:browse".into()],
        budget_usd: Some(1.0),
        max_depth: 0,
        iat: 1711100000,
        exp: 4711100000,
    };

    // Warmup
    for _ in 0..100 {
        let t = CompactToken::create(&claims, &kp).unwrap();
        let _ = CompactToken::verify(&t, &kp.public_key_bytes()).unwrap();
    }

    // Measure creation
    let mut create_times = Vec::new();
    let mut token = String::new();
    for _ in 0..iterations {
        let start = Instant::now();
        token = CompactToken::create(&claims, &kp).unwrap();
        create_times.push(start.elapsed().as_nanos() as f64 / 1_000_000.0);
    }

    // Measure verification
    let mut verify_times = Vec::new();
    let pubkey = kp.public_key_bytes();
    for _ in 0..iterations {
        let start = Instant::now();
        let _ = CompactToken::verify(&token, &pubkey).unwrap();
        verify_times.push(start.elapsed().as_nanos() as f64 / 1_000_000.0);
    }

    create_times.sort_by(|a, b| a.partial_cmp(b).unwrap());
    verify_times.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let create_mean: f64 = create_times.iter().sum::<f64>() / iterations as f64;
    let verify_mean: f64 = verify_times.iter().sum::<f64>() / iterations as f64;
    let token_size = token.len();

    println!("{{");
    println!("  \"language\": \"rust\",");
    println!("  \"mode\": \"compact\",");
    println!("  \"iterations\": {},", iterations);
    println!("  \"create_mean_ms\": {:.4},", create_mean);
    println!("  \"create_p50_ms\": {:.4},", create_times[iterations / 2]);
    println!("  \"create_p99_ms\": {:.4},", create_times[(0.99 * iterations as f64) as usize]);
    println!("  \"verify_mean_ms\": {:.4},", verify_mean);
    println!("  \"verify_p50_ms\": {:.4},", verify_times[iterations / 2]);
    println!("  \"verify_p99_ms\": {:.4},", verify_times[(0.99 * iterations as f64) as usize]);
    println!("  \"token_size_bytes\": {}", token_size);
    println!("}}");
}
